Feature/graphcast model

azadeh-gh opened the pull request:
### Refactor: Integrate Ocelot2 (Graphcast) Architecture and Refine Training Pipeline

**closes issue #35 and #36** 
**Summary**
This pull request introduces a significant architectural overhaul by integrating the Graphcast model into the Ocelot codebase. The core graph structure has been refactored from a homogeneous graph to a **heterogeneous graph** (`HeteroData`), enabling the model to handle diverse, multi-instrument observational data (e.g., satellite radiances and conventional surface data) in a modular and scalable way.

This update includes the adoption of Ocelot2 components along with critical improvements to the data pipeline, loss calculation, and distributed training stability.

### Key Changes and Refinements
**1. Core Architectural Overhaul (Aligning with Ocelot2)**

- **Data Representation (HeteroData):** The data processing pipeline in `gnn_datamodule.py` and `process_timeseries.py` has been completely refactored. Instead of creating a single large graph, it now constructs a `torch_geometric.HeteroData` object for each time step, with distinct node and edge types for each instrument.

- **Modular Model Architecture:** The `gnn_model.py` has been rewritten to adopt a modular ModuleDict structure. Encoders, decoders, and output mappers are now created dynamically based on the observation_config, making it easy to add or remove instruments.

- **GNN Layer Standardization:** The custom `EdgeGATLayer` has been replaced with the `Interaction_network` message-passing layer to be identical to the Ocelot2 implementation.

- **Dedicated Processor Module:** The main message-passing loop has been encapsulated into a dedicated Processor class, which correctly replicates the Ocelot2 architecture for deep, multi-step graph processing.

**2. Performance and Stability Improvements**

- **Memory Management:** Implemented gradient checkpointing within the Processor to dramatically reduce GPU memory usage, enabling the training of deeper models (e.g., 8+ layers) on high-resolution meshes without out-of-memory errors.
- **Distributed Training Fixes:**
- Resolved a critical deadlock issue in multi-node training by adding a `trainer.is_global_zero` guard to the data resampling callbacks, ensuring the expensive `datamodule.setup()` is only executed on the main process.

- Fixed a DDP RuntimeError by implementing the "dummy loss" trick, which ensures all model parameters are used in the backward pass, even when a batch is missing a specific observation type.

**3. Data Handling and Feature Engineering**

- **Generalized Scan Angle Embedding:** The logic for handling `scan_angle` has been generalized within the decoder, making it easy to apply to any satellite instrument, not just ATMS.
- **Data Subsampling:** Implemented a fixed-rate subsampling for observations ([::25] for satellite) to match the Ocelot2 data handling strategy, helping to manage graph size and improve performance.

**4. Tooling and Diagnostics**

- **Data Analysis Script:** Added a new diagnostic script, `analyze_zarr.py`, which provides statistical summaries and generates spatial distribution maps for any given Zarr dataset, improving our ability to inspect and validate input data.

**Ocelot2 Code Integration**
The following scripts were adapted directly from the Ocelot2 repository to serve as the foundational components for the new model architecture:

- `create_mesh_graph_global.py`
- `graphcast_aux.py`
- `grid_mesh_connectivity.py`
- `icosahedral_mesh.py`
- `interaction_net.py`
- `interaction_network.py`
- `model_utils.py`
- `processor.py`


- **Benchmark Plots:** benchmark plots from Epoch 70 of the model run to serve as a visual baseline for performance evaluation.
<img width="1600" height="500" alt="bt_map_channel_1 (4)" src="https://github.com/user-attachments/assets/d5b247cc-1b3f-488c-b381-e737a0f78081" />

<img width="1600" height="500" alt="bt_map_channel_2 (1)" src="https://github.com/user-attachments/assets/0b016e80-18de-4570-baf2-1834df2b1930" />
<img width="1600" height="500" alt="bt_map_channel_3 (1)" src="https://github.com/user-attachments/assets/6d26e6d3-f13c-4977-a0a6-c01edce56359" />
<img width="1600" height="500" alt="bt_map_channel_4 (1)" src="https://github.com/user-attachments/assets/ee9cb652-e232-4e65-9ced-f253e886a330" />
<img width="1600" height="500" alt="bt_map_channel_5 (1)" src="https://github.com/user-attachments/assets/48ff2034-e7a9-4eb4-b4a4-e8610912edf1" />
<img width="1600" height="500" alt="bt_map_channel_6" src="https://github.com/user-attachments/assets/bcfcc468-25b6-40f3-b345-70c53380d091" />

### Next Steps #16 
The immediate next step following this integration is to expand the model's capabilities by incorporating additional observation types from the v5 dataset located at `/scratch3/NCEPDEV/da/Ronald.McLaren/shared/ocelot/data_v5/global`. The focus will be on integrating variables that have Quality Control (QC) flags applied to further improve model accuracy and robustness.

azadeh-gh commented on gnn_model/create_mesh_graph_global.py:
```diff
@@ -0,0 +1,418 @@
+    parser.add_argument(
+        "--plot",
+        type=int,
+        default=0,
```
@CoryMartin-NOAA This particular script was adapted directly from the Ocelot2 repository, and I kept that argument as-is to maintain consistency with the original source code. However, we aren't actually using this command-line plotting feature since we've developed our own separate evaluation and plotting scripts.

xincjin-NOAA reviewed (dismissed):
Great job integrating Ocelot2 with Ocelot and achieving reasonable results. I’d prefer to approve this PR for now and leave some topics for discussion in our next meeting.

mlaurako commented on gnn_model/train_gnn.py:
```diff
@@ -1,123 +1,198 @@
+        "sensorZenithAngle": (31.15, 18.43),
+        "solarAzimuthAngle": (125.22, 95.04),
+        "solarZenithAngle": (71.94, 47.49),
+    }
```
How were these stat data pre-computed? Were they based on the v5 dataset?

xincjin-NOAA replied:
They were calculated using the v4 dataset, read from the Zarr file, converted to a Pandas DataFrame, and then used to compute the mean and standard deviation.

mlaurako commented on gnn_model/train_gnn.py:
```diff
@@ -1,123 +1,198 @@
         start_date=start_date,
         end_date=end_date,
         observation_config=observation_config,
+        mesh_structure=model.mesh_structure,  # This now works correctly
```
Perhaps clean up this comment?

mlaurako commented on gnn_model/run_gnn.sh:
```diff
@@ -41,5 +42,8 @@ echo "Visible GPUs on this node:"
 # Launch training
-srun --cpu_bind=cores python train_gnn.py --verbose
+# srun --cpu_bind=cores python train_gnn.py --verbose
+srun --cpu-bind=map_cpu:0,1,2,3 python train_gnn.py --verbose
```
is "--sampling_mode sequential" supposed to work in this version?

azadeh-gh commented on gnn_model/train_gnn.py:
```diff
@@ -155,27 +232,24 @@ def main():
-    elif args.sampling_mode == 'sequential':
+        callbacks.append(
+            ResampleDataCallback(
+                full_start_date="2024-04-01", full_end_date="2024-06-01", window_days=14
```
@CoryMartin-NOAA Done

azadeh-gh commented:
### New changes:

**- Config unification:**

  - Replaced configs/weights_config*.yaml with a single configs/observation_config.yaml.
  - New YAML schema: pipeline.subsample, instrument_weights, channel_weights, observation_config, and feature_stats.
  - Added optional qc_filters per instrument (e.g., pressure ranges).

**- Loss (loss.py):** Moved different weighted loss functions to the loss.py

**- DataModule (gnn_datamodule.py):** Builds instrument_ids tensors for each target node type.

**- Evaluation (evaluations.py):** 
  - Generalized plotting utility plot_instrument_maps(...).
  - Generates GT, prediction, and % error maps per channel; safer division and consistent colorbars.

**Training entrypoint (train.py)** 

- Switched to loading everything from configs/observation_config.yaml.
- Introduced rolling window training via callbacks (ResampleDataCallback / SequentialDataCallback) with FULL_START_DATE, FULL_END_DATE, WINDOW_DAYS.
- clearer prints, same DDP setup, and resume-from-checkpoint support preserved.

**Timeseries processing (process_timeseries.py)** 

- Subsampling rates now driven by config (pipeline.subsample).

- Added QC
- Replace Zarr fill values with NaN and drop invalid rows.
- Added time features (sin/cos of time-of-day) for inputs.
